modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockHttpProxyServer.java

```diff
@@ -7,91 +7,67 @@
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http.HttpServerCodec;
 
-import java.io.BufferedInputStream;
-import java.io.BufferedOutputStream;
 import java.io.Closeable;
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
-import java.net.Socket;
-import java.net.SocketException;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.net.InetSocketAddress;
+import java.util.function.Supplier;
 
 /**
  * A mock HTTP Proxy server for testing of support of HTTP proxies in various SDKs
  */
 class MockHttpProxyServer implements Closeable {
 
-    private static final Logger log = LogManager.getLogger(MockHttpProxyServer.class);
+    private final EventLoopGroup group = new NioEventLoopGroup(1);
 
-    private final MockServerSocket serverSocket;
-    private final Thread serverThread;
-    private final CountDownLatch latch;
-    private final ExecutorService executorService = Executors.newCachedThreadPool();
+    private ChannelFuture channelFuture;
+    private InetSocketAddress socketAddress;
 
-    MockHttpProxyServer(SocketRequestHandler handler) throws IOException {
-        // Emulate a proxy HTTP server with plain sockets because MockHttpServer doesn't work as a proxy
-        serverSocket = new MockServerSocket(0);
-        latch = new CountDownLatch(1);
-        serverThread = new Thread(() -> {
-            latch.countDown();
-            while (Thread.currentThread().isInterrupted() == false) {
-                Socket socket;
-                try {
-                    socket = serverSocket.accept();
-                } catch (SocketException e) {
-                    // Server socket is closed
-                    break;
-                } catch (IOException e) {
-                    log.error("Unable to accept socket request", e);
-                    break;
-                }
-                executorService.submit(() -> {
-                    try (
-                        socket;
-                        var is = new BufferedInputStream(socket.getInputStream());
-                        var os = new BufferedOutputStream(socket.getOutputStream())
-                    ) {
-                        // Don't handle keep-alive connections to keep things simple
-                        handler.handle(is, os);
-                    } catch (IOException e) {
-                        log.error("Unable to handle socket request", e);
+    MockHttpProxyServer handler(Supplier<SimpleChannelInboundHandler<FullHttpRequest>> handler) {
```

Comment:
Why do we need a `Supplier` here, can't we just pass the handler directly?

Reply:
If we pass the handler directly, then Netty complains that we add handlers add to pipelines multiple times without declaring them as `@Sharable`. But, `WebProxyServer`'s handler, is not sharable. So, we have to use fresh copies of handlers.

Reply:
Ah got it that makes sense.